pygfx/utils/transform.py

```diff
@@ -0,0 +1,555 @@
+    callbacks using the ``callback_id = transform.on_update(...)`` method and -
+    if the callee is a class - may optionally choose to decorate the callback
+    method with the `@callback` descriptor defined above. This will turn the
+    callback into a weakref and allow the callee to be garbage collected more
+    swiftly. After registration callees can remove the callback by calling
+    ``transform.remove_callback`` and passing the callback. If the callback was
+    a lambda, callees can also pass the ``callback_id`` returned when the
+    callback was first registered.
+
+    It also implements a basic caching mechanism that keeps computed properties
+    around until the underlying transform changes. This makes use of the
+    `@cached` descriptor defined above. The descriptor expects a `last_modified`
+    property on the consuming class which is used as a monotonously increasing
+    timestamp/counter to indicate if and when a cached value has become invalid.
+    Once invalid, cached values are updated upon the next read/get meaning that
+    they are updated lazily.
+
+    Parameters
+    ----------
+    gravity : ndarray, [3]
+        The direction of the gravity vector expressed in the target frame. It is
+        the inverse of ``WorldObject._up`` and used by the axis properties
+        (right, up, forward) to maintain a common level of rotation around an
+        axis when it is updated by it's setter. By default, it points along the
+        negative Y-axis.
+    is_camera_space : bool
+        If True, the transform represents a camera space which means that it's
+        ``forward`` and ``right`` directions are inverted.
+
+    Notes
+    -----
+    Subclasses need to define and implement ``last_modified`` for the caching
+    machnism to work correctly. Check out existing subclasses for an example of
+    how this might look like.
+
+    """
+
+    last_modified: int
+
+    def __init__(self, *, gravity=(0, -1, 0), is_camera_space=False):
+        self.update_callbacks = {}
+        self.is_camera_space = int(is_camera_space)
+        self._gravity = np.asarray(gravity, dtype=float)
+
+    @property
+    def matrix(self):
+        raise NotImplementedError()
+
+    @cached
+    def inverse_matrix(self) -> np.ndarray:
+        return np.linalg.inv(self.matrix)
+
+    @cached
+    def _decomposed(self) -> Tuple[np.ndarray, np.ndarray, np.ndarray]:
+        return la.matrix_decompose(self.matrix)
+
+    @cached
+    def _directions(self):
+        # Note: forward_is_minus_z indicates the camera frame
+        directions = (2 * self.is_camera_space - 1, 1, 1 - 2 * self.is_camera_space)
+
+        axes = np.diag(directions)
+        return (*la.vector_apply_matrix(axes, self.matrix),)
+
+    def flag_update(self):
+        for callback in self.update_callbacks.values():
+            callback(self)
+
+    def on_update(self, callback):
+        callback_id = id(callback)
+        self.update_callbacks[callback_id] = callback
```

Comment:
I'm also fine with putting this on the list of things to do in later PR's.

Reply:
> Can we also do the weakref trick here?

We are doing this the other way around by having callees register a `@callback` if they wish their lifecycle to not depend on the callback. This is more explicit (`@callback` clearly highlights what the method is meant to do) and it makes `AffineBase.on_update` much simpler.

Performance implications of wrapping everything in a `weakref` aside, it is actually not that trivial to decide which handles to wrap. If you wrap a `lambda`, it will get deallocated at the end of `AffineBase.on_update` as you lose the only hard reference to it. The same is - I think - true for `functools.partial`. The same scenario for nested functions, which will disappear at the end of the function in which they are defined. "Normal" Python functions (which are methods of the module, because everything is a class in Python, lol) are something you don't want to wrap, so you need to identify them and avoid wrapping them. You may be tempted to rely on `inspect.ismethod` to do this, but it only captures methods defined in pure python and will miss methods of built-in types (like `List.append`) or methods from extension modules (C-compiled python classes exposed in python). Then finally there are methods of nested classes, which look like an ordinary instance method to `on_update` (because they _are_), but they too will immediately deallocate once wrapped into a weakref, because the entire class definition is kept alive by a reference from its one instance, which in turn is kept alive by its one reference by the callback method, which we just wrapped into a weakref so the entire thing will get deallocated.

In sum, we can write logic that does the wrapping inside `on_update` and we can probably get it good enough to cover all the cases we care about, but the logic required to do this is (I think) too complex for the functionality it provides. Especially if a well-placed `@callback` can do the same job.

Ultimately, we introduced the `weakref` because GPU memory is precious and we want WorldObjects to be garbage collected asap instead of waiting for the GC to do its thing. Other objects typically don't hold references to such precious resources so it is fine to keep them around until the transform deallocates. CPython's GC can also deal with circular references on its own (no need for `weakref` here). It just doesn't do it at the end of each scope, which increases performance at the cost of memory usage - which is a bad tradeoff for WorldObject, hence `@callback`.

Reply:
Sorry, I could have been more clear. My idea would indeed be to *only* apply weakrefs to actual methods, which can AFAIK be detected reliably using `if hasattr(func, '__self__')`. But anyway, we could look into it later, or just leave it as is.

Reply:
> My idea would indeed be to only apply weakrefs to actual methods, which can AFAIK be detected reliably using if hasattr(func, '__self__')

Yes, the problem is that this might still do the wrong thing in some scenarios. For example, the following produces a dead weakref even though we probably wanted to have the callback (and the associated logger) survive until the transform goes away:

```python
import weakref

class StatefulDebugLogger:
    def __init__(self) -> None:
        self.event_count = 0

    def log_message(self):
        self.event_count += 1
        print(f"Message {self.event_count}: Something happened.")

def setup_logging():
    logger = StatefulDebugLogger()
    callback = logger.log_message

    assert hasattr(callback, "__self__"), "Not a method"
    # we are wrapping a method
    return weakref.ref(callback)

weak_callback = setup_logging()
assert weak_callback() is not None, "Callback is Dead"
```

We can make this work well for our common use cases, but getting this right for arbitrary callbacks is non-trivial.